File: week09/dfs.cpp
// C++ DFS 구현
#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

void dfs(int vertex, unordered_set<int>& visited, const vector<vector<int> >& graph) {
    visited.insert(vertex);
    cout << vertex << " ";
    for(auto neighbor : graph[vertex]) {
        if(visited.find(neighbor) == visited.end()) {
            dfs(neighbor, visited, graph);
        }
    }
}

int main() {
    // 그래프 인접 리스트 표현 (1-based indexing)
    int num_vertices = 4;
    vector<vector<int> > graph(num_vertices + 1);
    
    // 간선 추가 (무방향 그래프)
    graph[1].push_back(2);
    graph[1].push_back(4);
    graph[2].push_back(1);
    graph[2].push_back(3);
    graph[3].push_back(2);
    graph[3].push_back(4);
    graph[4].push_back(1);
    graph[4].push_back(3);
    
    unordered_set<int> visited;
    
    cout << "DFS 순회: ";
    dfs(1, visited, graph);
    cout << endl;
    
    return 0;
}

File: week09/dfs.py
# Python DFS 구현

def dfs(graph, start, visited=None):
    if visited is None:
        visited = set()
    visited.add(start)
    print(start, end=' ')
    for neighbor in graph[start]:
        if neighbor not in visited:
            dfs(graph, neighbor, visited)

if __name__ == "__main__":
    # 그래프 인접 리스트 표현
    graph = {
        1: [2, 4],
        2: [1, 3],
        3: [2, 4],
        4: [1, 3]
    }
    
    print("DFS 순회:")
    dfs(graph, 1)
    print()